[FX] Experimental type annotation pass using Python signatures

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53832 [FX] Normalize torch. namespace ops
* **#53831 [FX] Experimental type annotation pass using Python signatures**
* #53830 Function to retrieve inspect.Signature instances for PyTorch ops

Use Python function signatures to annotate types for Nodes within an FX graph.
This pulls out Python function signatures for both standard torch.nn Module classes
as well as `torch.nn.functional` functions. Support for `torch.` namespace ops
and tensor methods is forthcoming

Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

Commit:
[FX] Experimental type annotation pass using Python signatures

[ghstack-poisoned]

Commit:
Update on "[FX] Experimental type annotation pass using Python signatures"


Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

[ghstack-poisoned]

Commit:
Update on "[FX] Experimental type annotation pass using Python signatures"


Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

[ghstack-poisoned]

Commit:
Update on "[FX] Experimental type annotation pass using Python signatures"


Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

[ghstack-poisoned]

Commit:
Update on "[FX] Experimental type annotation pass using Python signatures"


Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

[ghstack-poisoned]

Commit:
Update on "[FX] Experimental type annotation pass using Python signatures"


Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

[ghstack-poisoned]

suo commented on torch/fx/experimental/schema_type_annotation.py:
```diff
@@ -0,0 +1,114 @@
+        """
+        assert not isinstance(target, str)
+        try:
+            sig = inspect.signature(inspect.unwrap(target))
```
I think by default `inspect.signature` follows wrapped callables, so the `inspect.unwrap` is not necessary.

suo commented on torch/fx/experimental/schema_type_annotation.py:
```diff
@@ -0,0 +1,114 @@
+            Optional[Dict]: Normalized kwargs for `target`, or `None` if this target is not
+                supported
+        """
+        assert not isinstance(target, str)
```
more appropriately `assert callable(target)` right?

suo commented on torch/fx/experimental/schema_type_annotation.py:
```diff
@@ -0,0 +1,114 @@
+
+    def _extract_python_return_type(self, target : Target) -> Optional[Dict[str, Any]]:
+        """
+        Given a call target, args, and kwargs, return the arguments normalized into
```
This docstring appears to be out of date. Also the return annotation.

suo commented on torch/fx/experimental/schema_type_annotation.py:
```diff
@@ -0,0 +1,114 @@
+            python_ret_type = self._extract_python_return_type(target_for_analysis)
+
+        return_proxy = super().call_function(target, args, kwargs)
+        return_proxy.node.type = return_proxy.node.type if return_proxy.node.type else python_ret_type
```
Out of curiosity: if `python_ret_type` was retrieved from a functional, under what conditions would it disagree with the proxy's type? Is it appropriate to always override the Python signature with the proxy type?

jamesr66a replied:
The `Proxy`'s type would have to have been set manually (none of our tools set it for anything other than placeholder and return), so we leave it alone so as to not disturb someone else's custom annotations

suo commented on torch/fx/experimental/schema_type_annotation.py:
```diff
@@ -0,0 +1,114 @@
+                    raise RuntimeError(f'Node referenced nonextent target {".".join(atoms[:i])}!')
+                module_itr = getattr(module_itr, atom)
+
+            maybe_inferred_ts_type = torch._C._jit_try_infer_type(module_itr)
```
Why this instead of `type(module_itr)`?

jamesr66a replied:
Because `type(module_itr)` will return types like `dict` rather than `Dict[K, V]` and that diverges from what the current convention is in FX

Commit:
Update on "[FX] Experimental type annotation pass using Python signatures"


Use Python function signatures to annotate types for Nodes within an FX graph.
This pulls out Python function signatures for both standard torch.nn Module classes
as well as `torch.nn.functional` functions. Support for `torch.` namespace ops
and tensor methods is forthcoming

Differential Revision: [D26982804](https://our.internmc.facebook.com/intern/diff/D26982804)

[ghstack-poisoned]
